### ISSUE TITLE
circuit breaker not generating failure rate exceeded event at expected interval

### ISSUE THREAD

**Issue by BigMikeATL** [resilience4j/resilience4j#2178]
Resilience4j version: 1.7.1

Java version: 8

I have r4j cb configured such that my expectation is that it will generate a failure rate exceeded event if the threshold of 100% is sustained for 3 minutes (180 seconds). That is not happening. Instead what I am seeing is that I will have some percentage of failures for about 2 minutes, it will jump to 100% sustained failures for about a minute, and then the failure rate exceeded event occurs.

Here is my configuration:
```
      slidingWindowSize: 180                  
      slidingWindowType: TIME_BASED           
      minimumNumberOfCalls: 10               
      failureRateThreshold: 100              
      eventConsumerBufferSize: 10
      recordExceptions:                       
        <recorded>
      ignoreExceptions:                      
        <ignored>
```

Is there an issue with the configuration? do I have a fundamental misunderstanding of the mechanics here? or do we have a bug?

I've attached a screenshot of the corresponding metrics, illustrating the behavior.
<img width="648" alt="Screenshot 2024-06-14 at 3 14 49 PM" src="https://github.com/resilience4j/resilience4j/assets/7064121/dfc72de8-62cd-4772-9695-cc7c7132f0da">

**Comment by BigMikeATL** [comment]
I'm seeing multiple fresh occurrences of this behavior in the last 72 hours. My expectation is that it takes the average failure rate of all snapshots for the last 180 seconds to determine whether the failure rate was exceeded.  Eyeballing the graph, this rate can't be but 60% or so, yet r4j says the rate was exceeded.

<img width="1222" alt="Screenshot 2024-06-17 at 9 46 05 AM" src="https://github.com/resilience4j/resilience4j/assets/7064121/acdd091c-9412-451d-b24f-dc46e156484c">

Jun 14, 2024 @ 07:15:16.449 | message:Circuit breaker failure rate exceeded. Event: 2024-06-14T14:15:16.449Z[GMT]: CircuitBreaker '2' exceeded failure rate threshold. Current failure rate: 100.0

Jun 14, 2024 @ 07:15:15.298 | message:Circuit breaker failure rate exceeded. Event: 2024-06-14T14:15:15.298Z[GMT]: CircuitBreaker '1' exceeded failure rate threshold. Current failure rate: 100.0

Note that the breaker is running in metrics only mode, but my expectation is that this mode behaves identically to regular mode with the only difference being that the circuit stays closed. Event generation behavior, such as failure rate being exceeded, should be identical in either case.

**Comment by RobWin** [comment]
Hi,

>I have r4j cb configured such that my expectation is that it will generate a failure rate exceeded event if the threshold of 100% is sustained for 3 minutes (180 seconds). 

No, this expectation is not correct. The sliding time window is 180 seconds long. Older call results are purged. `minimumNumberOfCalls` is 10, which means after 10 recorded calls the failure rate will be calculated. If all 10 calls have failed, the CircuitBreaker will transition to open. 
I tried to describe it here: https://resilience4j.readme.io/docs/circuitbreaker#time-based-sliding-window

This issue/thread might be interesting for you as well: https://github.com/resilience4j/resilience4j/issues/2173
MetricsOnly mode behaves different that the regular mode.

**Comment by BigMikeATL** [comment]
In regard to minimumNumberOfCalls, my interpretation is that we compute the failure rate of those 10 calls and aggregate them into the overall failure rate. And the value itself ensure we have _enough_ calls before recomputing as there would be limited value if we had, say, 2 calls in a minute. 

I had previously read the documentation about the time based sliding window prior to implementation and my interpretation of what's described is that we'd routinely recalculate and re-average the rates, and if we had a 100% failure rate for the last 180 seconds, the resulting rate is compared to the target rate, thus we have a rate exceeded event or we don't. In other words, it's a configurable (in this case, 180 second) rolling average. If it's not a rolling average, I guess I'm still not clear on what, exactly, the time based duration actually represents.

Based on the graphs, yes, there are 3 minutes with some degree of failures, but we only see a 100% failure rate for about a minute. I'm not sure why it's hanging out at 100% for a minute before deciding the failure rate has been exceeded.

If it's not too much trouble, is there a way to approximate the desired behavior? and/or can you explain in layman's terms how this time based configuration actually works, because it's not at all clear what the relationship between the time and failure rate is in relation to what's actually happening.

In regard to issue 2173, that's also my ticket and I do believe that's a bug as I see limited value in generating an event once and never again without programmatic intervention. As indicated in that ticket, I have a workaround in place where I reset the metrics and failure rate flag once the failure rate is exceeded.

**Comment by RobWin** [comment]
> In regard to minimumNumberOfCalls, my interpretation is that we compute the failure rate of those 10 calls and aggregate them into the overall failure rate. And the value itself ensures we have enough calls before recomputing as there would be limited value if we had, say, 2 calls in a minute.

The `minimumNumberOfCalls` parameter is only used for the initial phase of failure rate calculation in a circuit breaker. It specifies the minimum number of calls that must be made before the failure rate calculation starts. For example, if minimumNumberOfCalls is set to 10, the circuit breaker will not calculate the failure rate until it has monitored at least 10 calls. Once this threshold is reached, the failure rate is recalculated after each subsequent call.

Using a sliding window approach means that only the calls made within the last N seconds are considered for the current failure rate calculation. Calls that fall outside this time window are disregarded, ensuring that the failure rate reflects recent performance rather than being influenced by outdated data. 

It's important to note that there isn't an extra thread continuously recalculating the failure rate. Instead, the failure rate is recalculated immediately before a new call is permitted (or rejected). If this recalculation reveals that the failure rate exceeds the configured threshold, the circuit breaker transitions to the open state, preventing the call and thus protecting the system from further potential failures.

Regarding the metricsOnly mode, if your discussion or analysis is based on this mode, be aware that it behaves differently from the regular mode. The metricsOnly mode was implemented as a simpler, less accurate version due to time constraints and user requirements at the time. While it may not fully align with the expected behavior of a regular circuit breaker, it served its purpose for specific use cases. However, I agree that this implementation could be improved for better consistency and accuracy.


>In regard to issue 2173, that's also my ticket and I do believe that's a bug as I see limited value in generating an event once and never again without programmatic intervention. As indicated in that ticket, I have a workaround in place where I reset the metrics and failure rate flag once the failure rate is exceeded. 

Sorry, I missed it that it's you as well. To much work in parallel at work :(

**Comment by BigMikeATL** [comment]
@RobWin 
> 
> Using a sliding window approach means that only the calls made within the last N seconds are considered for the current failure rate calculation. Calls that fall outside this time window are disregarded, ensuring that the failure rate reflects recent performance rather than being influenced by outdated data.

The way I understand the above in regard to my specific configuration, it would mean that all the calls occurring in the last 180 second are considered for the current overall failure rate calculation. If you look at the graphs above, I'd reasonably expect that if what you said is true, we'd have an average of all successful vs failed calls for 180 second and r4j would conclude that we DO NOT have a 100% failure rate. This is evidenced by 2 minutes of stair stepping, with a mix if successes and failures, then it's only sustained at 100% for about a minute, at which point a failure rate exceeded event is emitted. So the behavior doesn't align with what I'd expect. That's where the hangup is.

Is the behavior because of insufficiencies in MetricsOnly mode? I've looked at the source but don't want to make any more potentially incorrect assumptions, preferring your expertise as the author to set me straight and advise on the best course of action. :)

If regular mode will behave the way I want due to insufficiencies in MetricsOnly mode, I can certain switch modes and make further tweaks.

**Comment by BigMikeATL** [comment]
@RobWin I changed the interval from 3 minutes to 4 minutes. Now I see it stair stepping longer, then it's only at 100% failure rate for a matter of seconds before the failure rate is deemed in excess.

<img width="964" alt="Screenshot 2024-06-28 at 4 28 13 PM" src="https://github.com/resilience4j/resilience4j/assets/7064121/bf66f708-810b-4ff3-bcfd-fac651f10753">

I don't understand how TIME_BASED, when set to 240 seconds with 100% failure rate, would consider the rate exceeded because the average of all success vs failed calls in that 240 second window doesn't come close to a 100% average. This behavior is unquestionably worse than when using a 180 second interval because in those cases at least we're seeing consistent behavior in terms of it stair stepping for 2 minutes, then maintaining a 100% failure rate for a minute before emitting the rate exceeded event. 

I'd assume metrics only mode uses the same exact failure rate calculation logic that the regular mode does. 

Either I have a really poor understanding of how TIME_BASED is supposed to behave, metrics only mode behaves differently than regular mode (haven't verified this yet), or there's an issue with how this works.

Bottom line is that if I set failure rate to 100% with TIME_BASED at X number of seconds, we haven't exceeded the failure rate until 100% of calls within the configured X number of seconds are failures.

**Comment by RobWin** [comment]
>Bottom line is that if I set failure rate to 100% with TIME_BASED at X number of seconds, we haven't exceeded the failure rate until 100% of calls within the configured X number of seconds are failures.

Yes, but that's the correct behavior.
If you have a sliding time window of X seconds and let's say 100 recorded calls, all of them must be failures so that the failure rate of the calls withint the sliding time window is 100%.

**Comment by BigMikeATL** [comment]
@RobWin This code is already in production and as you can see from the previous comments, the graphs and log entries clearly show the behavior. The behavior is unexpected. I'm not sure if/when I'll have time to write tests to simulate what I'm seeing in production but it's certainly doable, however I was hoping there'd be a clear explanation for the disconnect between expected and actual behavior.

My other option is to change production to not use metrics_only mode to see if behavior differs, but that adds risk as there are now significant side effects from an open circuit that I'd have to account for.

**Comment by BigMikeATL** [comment]
I reconfigured the service for 2 minutes @ 100% and unfortunately all we get is star stepping for a minute, and just like with 3 and 4 minutes, there's at best a sustained failure rate of 100% for about a minute before the failure rate is deemed in excess.

It's almost like the time based window setting doesn't truly matter, as if it evaluates only the last 60 seconds or so as part of the rate exceeded calculation, rather than looking at the average for the whole window.

<img width="1271" alt="Screenshot 2024-07-03 at 10 36 08 AM" src="https://github.com/resilience4j/resilience4j/assets/7064121/a111d8c3-2b0a-4d58-86ab-7af130f41464">
